externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/EngineEventsStore.scala

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.engine.spark.events
+
+import scala.collection.JavaConverters.collectionAsScalaIterableConverter
+
+import com.google.common.collect.TreeBasedTable
+
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.config.KyuubiConf.ENGINE_UI_SESSION_LIMIT
+
+/**
+ * A memory store that tracking the number of statements and sessions, it provides:
+ *
+ * - stores all events, and sorted by startTimestamp.
+ * - cleanup the last events when reach a certain threshold.
+ */
+class EngineEventsStore(conf: KyuubiConf) {
+
+  private type Long = java.lang.Long
+
+  /**
+   * The number of SQL client sessions kept in the Kyuubi Query Engine web UI.
+   */
+  private val retainedSessions: Int = conf.get(ENGINE_UI_SESSION_LIMIT)
+
+  /**
+   * store all session events, and sorted by startTimestamp.
+   */
+  val sessions: TreeBasedTable[Long, String, SessionEvent] = TreeBasedTable.create()
+
+  def getSessionList: Seq[SessionEvent] = {
+    sessions.values().asScala.toSeq
+  }
+
+  def getSession(sessionId: String): Option[SessionEvent] = {
+    sessions.values().asScala.find(s => s.sessionId == sessionId)
+  }
+
+  /**
+   * save session events and check the capacity threshold
+   */
+  def saveSession(sessionEvent: SessionEvent): Unit = synchronized {
+    sessions.put(sessionEvent.startTime, sessionEvent.sessionId, sessionEvent)
```

Comment:
shall we use endTime?

Reply:
use startTime is more resonable.

Reply:
can you tell me the reason then？ it seem to me that we very likely to keep late started but closed sessions，in the meanwhile，drop those still alive？

Reply:
the origin design about the retain may like a fixed size queue. if submit 300 queries at the same time, and the retain size is 200, no matter whether starttime or endtime is used, there will also be 100 active queries are excluded.

Reply:
if sort the events order by starttime, seems more easier to understand.

Reply:
can we always keep the live sessions?

Reply:
ok, will updated.